File: backend/DentalClinic.Core/Domain/Entities/Prescription.cs
namespace DentalClinic.Core.Domain.Entities;

public sealed class Prescription : Entity
{
    public Guid PatientId { get; private set; }
    public Patient Patient { get; private set; } = null!;

    public Guid DoctorId { get; private set; }
    public User Doctor { get; private set; } = null!;

    public string? Observations { get; private set; }
    public List<PrescriptionItem> Items { get; private set; } = new();

    public Guid ClinicId { get; private set; }

    private Prescription() { }

    public static Prescription Create(Guid patientId, Guid doctorId, Guid clinicId, string? observations)
    {
        return new Prescription
        {
            PatientId = patientId,
            DoctorId = doctorId,
            ClinicId = clinicId,
            Observations = observations
        };
    }

    public void AddItem(string medicine, string dosage, string instructions)
    {
        Items.Add(new PrescriptionItem(medicine, dosage, instructions));
    }
}

public record PrescriptionItem(string MedicineName, string Dosage, string Instructions);

File: lib/features/prontuario/domain/entities/prescription.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'prescription.freezed.dart';
part 'prescription.g.dart';

@freezed
class Prescription with _$Prescription {
  const factory Prescription({
    required String id,
    required String patientId,
    required String doctorId,
    required String doctorName,
    required DateTime date,
    required List<PrescriptionItem> items,
    String? observations,
    required String clinicId,
  }) = _Prescription;

  factory Prescription.fromJson(Map<String, dynamic> json) => _$PrescriptionFromJson(json);
  
  Map<String, dynamic> toJson() => _$$PrescriptionImplToJson(this as _$PrescriptionImpl);
}

@freezed
class PrescriptionItem with _$PrescriptionItem {
  const factory PrescriptionItem({
    required String medicineName,
    required String dosage,
    required String instructions,
  }) = _PrescriptionItem;

  factory PrescriptionItem.fromJson(Map<String, dynamic> json) => _$PrescriptionItemFromJson(json);
  
  Map<String, dynamic> toJson() => _$$PrescriptionItemImplToJson(this as _$PrescriptionItemImpl);
}

@freezed
class MedicalCertificate with _$MedicalCertificate {
  const factory MedicalCertificate({
    required String id,
    required String patientId,
    required String doctorId,
    required String doctorName,
    required DateTime date,
    required String content,
    required int daysOfRest,
    required String cid,
    required String clinicId,
  }) = _MedicalCertificate;

  factory MedicalCertificate.fromJson(Map<String, dynamic> json) => _$MedicalCertificateFromJson(json);

  Map<String, dynamic> toJson() => _$$MedicalCertificateImplToJson(this as _$MedicalCertificateImpl);
}
